### PR TITLE
Event pusher push RDBMS backend

### PR DESCRIPTION
**Pull request by michalwski**
This PR adds RDBMS backend for mod_evet_pusher_push

Proposed changes include:
* Postgres, MySQL and MSSQL schemas
* RDBMS backend implementatino
* Updated tests to use RDBMS backed for push if possible

### CODE REVIEW

**Comment by mongoose-im** [comment]
[7430.1](https://travis-ci.org/esl/MongooseIM/jobs/608376356) / Erlang 22.0 / small_tests / a5ae000355308b217cae63bb42135afd0124d207
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2526/7430/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2526/7430/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-11-06_19.46.26/lib.mongooseim.logs/run.2019-11-06_19.46.26/suite.log.html)

File: src/event_pusher/mod_event_pusher_push_rdbms.erl [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    {error, Reason :: term()}.
+get_publish_services(#jid{lserver = LServer} = User) ->
+    UserEscaped = mongoose_rdbms:escape_string(jid:to_binary(jid:to_lus(User))),
+    SelectQuery = select_query(UserEscaped),
+    Result = mongoose_rdbms:sql_query(LServer, SelectQuery),
+    case Result of
+        {selected, Rows} ->
+            {ok, decode_rows(Rows)};
+        Other ->
+            Other
```
Comment (aleklisi) [inline src/event_pusher/mod_event_pusher_push_rdbms.erl]: log the 'Other` case?

File: src/event_pusher/mod_event_pusher_push_rdbms.erl [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    ok | {error, Reason :: term()}.
+disable(#jid{lserver = LServer} = User, PubSub, Node) ->
+    UserEscaped = mongoose_rdbms:escape_string(jid:to_binary(jid:to_lus(User))),
+    PubSubBin = maybe_to_binary(PubSub),
+    PubSubEscaped = escape_if_defined(PubSubBin),
+    NodeEscaped = escape_if_defined(Node),
+    DeleteQuery = delete_query(UserEscaped, PubSubEscaped, NodeEscaped),
+    case mongoose_rdbms:sql_query(LServer, DeleteQuery) of
+        {updated, _} -> ok;
+        Other -> Other
```
Comment (aleklisi) [inline src/event_pusher/mod_event_pusher_push_rdbms.erl]: log the 'Other` case?

File: src/event_pusher/mod_event_pusher_push_rdbms.erl [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+decode_rows(Rows) ->
+    [decode_row(Row) || Row <- Rows].
+
+decode_row({NodeID, PubSubBin, FormJSON}) ->
+    {jid:from_binary(PubSubBin),
+     NodeID,
+     decode_form(FormJSON)}.
+
+insert_query(User, PubSub, Node, FormJSON, CreatedAt) ->
```
Comment (aleklisi) [inline src/event_pusher/mod_event_pusher_push_rdbms.erl]: how about:
``` erlang
insert_query(RawValues = [_User, _PubSub, _Node, _FormJSON, _CreatedAt]) ->
  EscValues = lists:map(fun mongoose_rdbms:use_escaped_string/1, RawValues),
  Values = string:join(EscValues, ", "),
  [
    <<"INSERT INTO ">>, table_name(),
    <<" (owner_jid, node, pubsub_jid, form, created_at) VALUES (">>,
    Values,  ")"
  ].
```
Is is more readable? If not then current version is also good :)

File: src/event_pusher/mod_event_pusher_push_rdbms.erl [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    DeleteQuery = delete_query(UserEscaped, PubSubEscaped, NodeEscaped),
+    mongoose_rdbms:sql_query(LServer, DeleteQuery),
+
+    CreatedAtEscaped = mongoose_rdbms:escape_integer(os:system_time(microsecond)),
+    InsertQuery = insert_query(UserEscaped, PubSubEscaped, NodeEscaped,
+                               FormsJSONEscaped, CreatedAtEscaped),
+    case mongoose_rdbms:sql_query(LServer, InsertQuery) of
+        {updated, 1} -> ok;
+        Other -> Other
```
[the remark this answers is not in the document]
Comment (michalwski) [reply src/event_pusher/mod_event_pusher_push_rdbms.erl]: I think there's no point in logging it here. It's handled by caller of backend functions.

File: src/event_pusher/mod_event_pusher_push_rdbms.erl (hunk 1 above)
Comment (michalwski) [reply src/event_pusher/mod_event_pusher_push_rdbms.erl]: Same as for the the other `Other` :)

File: src/event_pusher/mod_event_pusher_push_rdbms.erl (hunk 2 above)
Comment (michalwski) [reply src/event_pusher/mod_event_pusher_push_rdbms.erl]: no? :)

File: src/event_pusher/mod_event_pusher_push_rdbms.erl (hunk 3 above)
Comment (michalwski) [reply src/event_pusher/mod_event_pusher_push_rdbms.erl]: It is not more readable to me. More over, notice that `CreatedAt` is an escaped integer and not string :)

**Comment by mongoose-im** [comment]
[7433.1](https://travis-ci.org/esl/MongooseIM/jobs/608665446) / Erlang 22.0 / small_tests / 28999f05a7f30a8b7f09740ac0ea535c2c9b8772
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2526/7433/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2526/7433/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-11-07_11.23.21/lib.mongooseim.logs/run.2019-11-07_11.23.22/suite.log.html)

**Pull request merged**